app/scripts/controllers/detect-tokens.js

```diff
@@ -79,7 +80,10 @@ export default class DetectTokensController {
     }
 
     const { tokenList } = this._tokenList.state;
-    if (Object.keys(tokenList).length === 0) {
```

Comment:
Do we want to get rid of this check?

Reply:
Even though the tokenList will be empty for testnets when the usetokenDetection flag is ON when the flag is OFF the token List will be holding the static tokenList which mimics the use of contract-metadata before the introduction of token detection. So the above test was failing and the token detection was getting triggered while on testnets. 

The token detection is currently available only on Mainnet, so the new check will make sure the detection will be triggered only for Mainnet. Once we have the token detection enabled in other networks we will need to update the current check I have added to make sure token detection will work on all those networks.

We can also do something like this, but I am wondering if this is worth it considering the current scenario.

    if (this._network.store.getState().provider.chainId !== MAINNET_CHAIN_ID && (!this.useTokenDetection || Object.keys(tokenList).length === 0)) {
      return;
    }

Reply:
Can we be quite confident that tokenList will never have a length of 0 for Mainnet (or any other network for which we might enable token detection down the line)?

Reply:
If tokenList length is 0, the token detection will not work as we are using the token address from tokenList for balance check. When token detection is ON, as of now the testnets will have an empty token list. When token detection is OFF the tokenList will have the data from contract-metadata. If tokenList is empty for any network the token detection will be disabled but  we can always add any token deployed in a network via `custom token`

Reply:
I think the alternative suggested in this comment is good: https://github.com/MetaMask/metamask-extension/pull/12232/files#r717893459